### PR TITLE
Feature: add FunctionDecl and CompoundStmt in AST

### LINKED ISSUE #52: Implement `gil::ast::FunctionDecl` class

### PR DESCRIPTION
**Pull request by LindonAliu**
fix #52 #147

### CODE REVIEW

File: include/AST/Param.hpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#ifndef GLU_AST_PARAM_HPP
+#define GLU_AST_PARAM_HPP
+
+#include "Types/TypeBase.hpp"
+#include <string>
+
+namespace glu::ast {
+
+struct Param {
```
Comment (Snowy1803) [inline include/AST/Param.hpp:9]: you can put this directly in the FunctionDecl class

File: include/AST/FunctionDecl.hpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
+#ifndef GLU_AST_DECL_FUNCTIONDECL_HPP
+#define GLU_AST_DECL_FUNCTIONDECL_HPP
+
+#include "ASTNode.hpp"
+#include "Types/Types.hpp"
+
+#include "llvm/ADT/ArrayRef.h"
```
Comment (Snowy1803) [inline include/AST/FunctionDecl.hpp:7]: <>

File: include/AST/FunctionDecl.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    std::string getName() const { return _name; }
+
+    /// @brief Getter for the type of the function.
+    /// @return Returns the type of the function.
+    glu::types::FunctionTy *getType() const { return _type; }
+
+    /// @brief Getter for the parameters of the function.
+    /// @return Returns a vector of Param objects representing the parameters of
+    /// the function.
+    llvm::ArrayRef<Param> const &getParams() const { return _params; }
```
Comment (Snowy1803) [inline include/AST/FunctionDecl.hpp:65]: ArrayRef is already a ref(erence)
```suggestion
    llvm::ArrayRef<Param> getParams() const { return _params; }
```

File: include/AST/FunctionDecl.hpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+/// @class FunctionDecl
+/// @brief Represents a function declaration in the AST.
+///
+/// This class inherits from DeclBase and encapsulates the details of a function
+/// declaration, including its name, type, parameters, and body.
+class FunctionDecl : public DeclBase {
+    std::string _name;
+    glu::types::FunctionTy *_type;
+    llvm::SmallVector<Param> _params;
+    glu::ast::ASTNode *_body;
```
Comment (Snowy1803) [inline include/AST/FunctionDecl.hpp:33]: the body should be a compound statement

File: test/AST/FunctionDecl.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+TEST_F(FunctionDeclTest, FunctionDeclConstructor)
+{
+    std::string name = "foo";
+    std::vector<glu::types::TypeBase *> parameters
+        = { new glu::types::BoolTy(), new glu::types::BoolTy() };
+    glu::types::TypeBase *returnType = new glu::types::BoolTy();
+
+    glu::types::FunctionTy type(parameters, returnType);
+    llvm::SmallVector<Param> params = { Param("a", new glu::types::BoolTy()),
+                                        Param("b", new glu::types::BoolTy()) };
```
Comment (Snowy1803) [inline test/AST/FunctionDecl.cpp:30]: don't recreate new BoolTy every time, use the same instance (in the real implementation, all types will be uniqued)

File: test/AST/FunctionDecl.cpp [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    ASSERT_TRUE(llvm::isa<FunctionDecl>(test));
+    ASSERT_FALSE(llvm::isa<StmtBase>(&decl));
+
+    for (auto param : decl.getParams()) {
+        delete param.type;
+    }
+    delete returnType;
+    for (auto param : parameters) {
+        delete param;
+    }
```
Comment (Snowy1803) [inline test/AST/FunctionDecl.cpp:47]: should only be one type to delete, then

**Comment by LindonAliu** [comment]
setParent in CompoundStmt is in #144 PR

File: include/AST/Stmt/CompoundStmt.hpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        llvm::SmallVector<StmtBase *> stmts
+    )
+        : StmtBase(NodeKind::CompoundStmtKind, location, parent)
+        , _stmts(std::move(stmts))
+    {
+    }
+
+    /// @brief Get the list of statements in the compound statement.
+    /// @return A reference to the list of statements.
+    llvm::ArrayRef<StmtBase *> getStmts() const { return _stmts; }
```
Comment (Snowy1803) [inline include/AST/Stmt/CompoundStmt.hpp:36]: const alternatives are useless and to be avoided

**Review by Snowy1803 — APPROVED** [review]

**Review by jeremyelalouf — APPROVED** [review]

**Review by yunsecode — APPROVED** [review]

**Review by Snowy1803 — APPROVED** [review]
LGTM

**Review by jeremyelalouf — APPROVED** [review]

**Pull request merged**